code/go/0chain.net/blobbercore/challenge/challenge.go

```diff
@@ -135,49 +137,112 @@ func processAccepted(ctx context.Context) {
-		for _, c := range challenges {
-			logging.Logger.Info("[challenge]process: ",
-				zap.String("challenge_id", c.ChallengeID),
-				zap.Time("created", c.CreatedAt))
-			err := c.UnmarshalFields()
-			if err != nil {
-				logging.Logger.Error("[challenge]process: ",
-					zap.String("challenge_id", c.ChallengeID),
-					zap.Time("created", c.CreatedAt),
-					zap.String("validators", string(c.ValidatorsString)),
-					zap.String("lastCommitTxnList", string(c.LastCommitTxnList)),
-					zap.String("validationTickets", string(c.ValidationTicketsString)),
-					zap.String("ObjectPath", string(c.ObjectPathString)),
-					zap.Error(err))
-				continue
-			}
-			swg.Add()
-			go validateChallenge(&swg, c)
+
+	rows, err := db.Model(&ChallengeEntity{}).
+		Where("status = ?", Accepted).
+		Select("challenge_id", "created_at").Rows()
+
+	if err != nil {
+		logging.Logger.Error("[challenge]process: ",
+			zap.Error(err))
+		return
+	}
+
+	defer rows.Close()
+
+	startTime := time.Now()
+	swg := sizedwaitgroup.New(config.Configuration.ChallengeResolveNumWorkers)
+	count := 0
+	for rows.Next() {
+		count++
+		now := time.Now()
+
+		var challengeID string
+		var createdAt time.Time
+
+		err := rows.Scan(&challengeID, &createdAt)
+		if err != nil {
+			logging.Logger.Error("[challenge]process: ",
+				zap.Error(err))
+			continue
 		}
-		swg.Wait()
 
-		logging.Logger.Info("[challenge]elapsed:process ",
-			zap.Int("count", len(challenges)),
-			zap.String("save", time.Since(startTime).String()))
+		if time.Since(createdAt) > config.Configuration.ChallengeCompletionTime {
+
+			db.Model(&ChallengeEntity{}).
+				Where("challenge_id =? and status =? ", challengeID, Accepted).
+				Updates(map[string]interface{}{
+					"status":         Cancelled,
+					"result":         ChallengeFailure,
+					"status_message": fmt.Sprintf("created: %s, start: %s , delay: %s, cct: %s", createdAt, now, now.Sub(createdAt).String(), config.Configuration.ChallengeCompletionTime.String()),
+				})
+
+			logging.Logger.Error("[challenge]process: timeout ",
+				zap.Any("challenge_id", challengeID),
+				zap.Time("created", createdAt),
+				zap.Time("start", now),
+				zap.String("delay", now.Sub(createdAt).String()),
+				zap.String("cct", config.Configuration.ChallengeCompletionTime.String()),
+				zap.Error(err))
+			continue
+		}
+
+		swg.Add()
+		go func(id string) {
+			defer swg.Done()
+			validateChallenge(id)
+		}(challengeID)
+
 	}
-}
 
-func validateChallenge(swg *sizedwaitgroup.SizedWaitGroup, c *ChallengeEntity) {
-	defer swg.Done()
+	swg.Wait()
+
+	logging.Logger.Info("[challenge]elapsed:process ",
+		zap.Int("count", count),
+		zap.String("save", time.Since(startTime).String()))
+
+}
 
+func validateChallenge(id string) {
 	startTime := time.Now()
 
 	ctx := datastore.GetStore().CreateTransaction(context.TODO())
 	defer ctx.Done()
 
+	var c *ChallengeEntity
+
 	db := datastore.GetStore().GetTransaction(ctx)
```

Comment:
if it's a transaction we should use `tx`
if database `db`
makes more readable..

Reply:
fixed